elki/src/main/java/elki/math/statistics/dependence/MCDEDependenceMeasure.java

```diff
@@ -0,0 +1,249 @@
+    bibkey = "DBLP:conf/ssdbm/FoucheB19")
+
+public class MCDEDependenceMeasure
+    extends AbstractDependenceMeasure {
+  /**
+   * Monte-Carlo iterations.
+   */
+  protected int m = 50;
+
+  /**
+   * Expected share of instances in slice (independent dimensions).
+   */
+  protected double alpha = 0.5;
+
+  /**
+   * Parameter that specifies the size of the marginal restriction. Note that in the original paper
+   * alpha = beta and as such there is no explicit distinction between the parameters.
+   */
+  protected double beta = 0.5;
+
+  /**
+   * Random generator.
+   */
+  protected RandomFactory rnd;
+
+  /**
+   * Statistical Test returning p-value tailored to MCDE Framework.
+   */
+  protected MCDETest mcdeTest;
+
+  /**
+   * Constructor.
+   */
+  public MCDEDependenceMeasure(int m, double alpha, double beta, RandomFactory rnd, MCDETest mcdeTest) {
+    this.m = m;
+    this.alpha = alpha;
+    this.beta = beta;
+    this.rnd = rnd;
+    this.mcdeTest = mcdeTest;
+  }
+
+  /**
+   * Data Slicing
+   *
+   * @param len         No of data instances
+   * @param nonRefIndex Index (see correctedRank()) computed for the dimension that is not the reference dimension
+   * @return Array of booleans that states which instances are part of the slice
+   */
+  protected boolean[] randomSlice(int len, MCDETest.RankStruct[] nonRefIndex) {
+    final Random random = rnd.getSingleThreadedRandom();
+    boolean slice[] = new boolean[len];
+    Arrays.fill(slice, true);
+
+    // According to the actual formula it should be Math.ceil(Math.pow(this.alpha, 1.0) * len).
+    // The exponent in the multivariate case should be the no. of dimensions - 1 which in the
+    // bivariate case is always simply 1.
+    final int slizeSize = (int) Math.ceil(this.alpha * len);
+    final int start = random.nextInt(len - slizeSize);
+    final int end = start + slizeSize;
+
+    for(int j = 0; j < start; j++) {
+      slice[nonRefIndex[j].index] = false;
+    }
+
+    for(int j = end; j < len; j++) {
+      slice[ nonRefIndex[j].index] = false;
+    }
+
+    return slice;
+  }
+
+  /**
+   * Implements dependence from DependenceMeasure superclass. Corresponds to Algorithm 4 in source paper.
+   *
+   * @param adapter1 First data adapter
+   * @param data1    First data set
+   * @param adapter2 Second data adapter
+   * @param data2    Second data set
+   * @param <A>      Numeric data type, such as double
+   * @param <B>      Numeric data type, such as double
+   * @return MCDE result
+   */
+  @Override public <A, B> double dependence(final NumberArrayAdapter<?, A> adapter1, final A data1, final NumberArrayAdapter<?, B> adapter2, final B data2) {
+    final Random random = rnd.getSingleThreadedRandom();
+    final int len = adapter1.size(data1);
+
+    if(len != adapter2.size(data2)) {
+      throw new AbortException("Size of both arrays must match!");
+    }
+
+    final MCDETest.RankStruct[] index_0 = mcdeTest.corrected_ranks(adapter1, data1, len);
```

Comment:
This array type is likely to break in the future. An `ArrayList<MCDETest.RankStruct>` is more tolerant, because then the actual type is erased in the compiled classes - its then just an `ArrayList` eventually.
Now if `corrected_ranks` would return a `MWPTest.MwpIndex[]`, this could (should?) cause an error, because of type covariance. The warning caused by the missing generics is actually hinting at such a problem, and I am actually surprised that it works.